### PR TITLE
Bugfix/1527 fix grad norm tracking

### LINKED ISSUE #1527: Track_grad_norm only tracks the parameters of the last optimizer defined
## 🐛 Bug

When you enable track_grad_norm in Trainer you expect it to track the grad of all the parameters defined in your lightning module. It seems like it only tracks the parameters for the last optimizer defined in `def configure_optimizers()`.  

### To Reproduce

1. Create a pytorch lightning module with more than one optimizer
2. Enable logging to tensorboard and track_grad_norm in the trainer
3. Train model and observe only gradients for the last optimzer is tracked
4. Delete all tensorboard data and restart tensorboard
5. Train a model with the optimizers returned in the different order. Observe which gradients are tracked again

#### Code sample

```
import pytorch_lightning as pl
from torch.utils.data import TensorDataset, DataLoader
from pytorch_lightning.loggers.tensorboard import TensorBoardLogger
from torch.optim import SGD
import torch.nn as nn
import torch

class MWENet(pl.LightningModule):
    def __init__(self):
        super(MWENet, self).__init__()
        
        self.first = nn.Conv2d(1, 1, 3)
        self.second = nn.Conv2d(1, 1, 3)
        self.loss = nn.L1Loss()
        
    def train_dataloader(self):
        xs, ys = torch.zeros(16, 1, 10, 10), torch.ones(16 ,1, 6, 6)
        ds = TensorDataset(xs, ys)
        return DataLoader(ds)
    
    def forward(self, xs):
        out = self.first(xs)
        out = self.second(out)
        return out
    
    def configure_optimizers(self):
        first = SGD(self.first.parameters(), lr=0.01)
        second = SGD(self.second.parameters(), lr=0.01)
        return [second, first]
    
    def training_step(self, batch, batch_idx, optimizer_idx):
        xs, ys = batch
        out = self.forward(xs)
        return {'loss': self.loss(out, ys)}

net = MWENet()
logger = TensorBoardLogger('somedir', name='testing')
trainer = pl.Trainer(track_grad_norm=2,
                     logger = logger)
trainer.fit(net)
```

### Expected behavior

I expect that a pytorch module will report the gradient of all parameters when track_grad_norm is enabled.

### Environment

```
cuda:
	GPU:
		GeForce RTX 2080 Ti
	available:           True
	version:             10.1
packages:
	numpy:               1.18.1
	pyTorch_debug:       False
	pyTorch_version:     1.4.0
	pytorch-lightning:   0.7.3
	tensorboard:         2.2.1
	tqdm:                4.45.0
system:
	OS:                  Windows
	architecture:
		64bit
		WindowsPE
	processor:           Intel64 Family 6 Model 158 Stepping 12, GenuineIntel
	python:              3.7.7
	version:             10.0.18362
```

### PR DESCRIPTION
**Pull request by tchaton**
fixes #1527

### CODE REVIEW

File: pytorch_lightning/trainer/supporters.py [hunk 1]
```diff
@@ -177,3 +178,94 @@ def to_disk(self) -> None:
             # Write predictions for current file to disk
             with fs.open(filepath, "wb") as fp:
                 torch.save(outputs, fp)
+
+
+class AverageValueMeter:
```
Comment (tchaton) [inline pytorch_lightning/trainer/supporters.py:183]: Can you make a reference to torchnet ?

File: pytorch_lightning/trainer/supporters.py [hunk 2]
```diff
@@ -177,3 +178,94 @@ def to_disk(self) -> None:
...
+        self.val = 0.0
+        self.mean = np.nan
+        self.mean_old = 0.0
+        self.m_s = 0.0
+        self.std = np.nan
+
+
+class GradNormTracker:
+
+    def __init__(self, aggregation_mode, norm_type):
```
Comment (tchaton) [inline pytorch_lightning/trainer/supporters.py:224]: Can you add docstring ?

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -261,6 +262,9 @@ def __init__(
 
             track_grad_norm: -1 no tracking. Otherwise tracks that p-norm. May be set to 'inf' infinity-norm.
 
+            track_grad_norm_mode: How grad norms are aggregated. By default it aggregates for each parameter.
+                Additionally it can aggregate for parameter+optimizer and optimizer.
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:266]:
```suggestion
                The possible options are `parameters`, `optimizer`, `optimizer+parameters`.
```

File: pytorch_lightning/trainer/supporters.py [hunk 4]
```diff
@@ -177,3 +178,94 @@ def to_disk(self) -> None:
...
+        self._grad_norm_dic = {}
+        self.norm_type = norm_type
+
+        if aggregation_mode == 'optimizer':
+            self.name_mapping = self.aggregate_over_optimizer
+        elif aggregation_mode == 'optimizer+parameters':
+            self.name_mapping = self.aggregate_over_optimizer_and_parameters
+        elif aggregation_mode =='parameters':
+            self.name_mapping = self.aggregate_over_parameters
```
Comment (tchaton) [inline pytorch_lightning/trainer/supporters.py:233]: Can you raise an exception and provide the correct option list if aggregation_mode is not one of them.

File: pytorch_lightning/trainer/connectors/training_trick_connector.py [hunk 5]
```diff
@@ -39,6 +40,12 @@ def on_trainer_init(
             raise MisconfigurationException("track_grad_norm can be an int, a float or 'inf' (infinity norm).")
         self.trainer.track_grad_norm = float(track_grad_norm)
 
+        # gradient norm tracking aggregation mode
+        if not isinstance(track_grad_norm_mode, str) and\
+           not track_grad_norm_mode in ['parameters', 'optimizer', 'optimizer+parameters']:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/training_trick_connector.py:45]: Better doing GradNormTracker.check_grad_norm_mode(track_grad_norm_mode) -> True or False.

File: tests/trainer/logging_tests/test_grad_norm_track.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+def test_grad_norm_track_aggregated_over_optimizers():
+    model = TestModel()
+    trainer = Trainer(track_grad_norm=2,
+                      track_grad_norm_mode='optimizer',
+                      log_every_n_steps=1,
+                      fast_dev_run=True)
+
+    trainer.fit(model)
+
+    first_optimizer = ('opt_0_grad_2.0_norm_mean',
```
Comment (tchaton) [inline tests/trainer/logging_tests/test_grad_norm_track.py:109]: in Optimizer mode, norm and norm_total should be the same. Duplicated information

Comment (tchaton) [reply tests/trainer/logging_tests/test_grad_norm_track.py:109]: And not sure if `std` makes sense.

File: tests/trainer/logging_tests/test_grad_norm_track.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                      fast_dev_run=True)
+
+    trainer.fit(model)
+
+    first_optimizer = ('grad_2.0_norm_first.weight_mean',
+                       'grad_2.0_norm_first.weight_std',
+                       'grad_2.0_norm_first.bias_mean',
+                       'grad_2.0_norm_first.bias_std',
+                       'grad_2.0_norm_total_mean',
+                       'grad_2.0_norm_total_std')
```
Comment (tchaton) [inline tests/trainer/logging_tests/test_grad_norm_track.py:64]: Not sure `total_std` really makes sense.

**Review by tchaton — COMMENTED** [review]
Great work there ! Minors changes.

File: tests/trainer/logging_tests/test_grad_norm_track.py (hunk 7 above)
Comment (MikkelAntonsen) [reply tests/trainer/logging_tests/test_grad_norm_track.py:64]: Also, I'm not quite sure if having std adds value. From experience, adding grad norm tracking adds an enormous amount of information in tensorboard. Having std doubles that.
I can remove std for total at least. Do we want std for the rest?

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (MikkelAntonsen) [reply pytorch_lightning/trainer/trainer.py:266]: Do you have a suggestion for better naming here? `optimizer+parameters` in particular is not a good name from a user perspective.

File: tests/trainer/logging_tests/test_grad_norm_track.py (hunk 6 above)
Comment (MikkelAntonsen) [reply tests/trainer/logging_tests/test_grad_norm_track.py:109]: > in Optimizer mode, norm and norm_total should be the same. Duplicated information

Not necessarily, as the code is today, I think. If you have a dense layer with weights and bias and you aggregate over an optimizer you will get `norm_mean = mean(weight.norm, bias.norm)` calculated by grad_norm_tracker. While `norm_total=mean( (weight+bias).norm )`, where the `(weight + bias).norm` is calculated in  GradInformation.py. 

In my opinion, the most natural thing to do would be norm_total.

File: pytorch_lightning/trainer/supporters.py (hunk 1 above)
Comment (MikkelAntonsen) [reply pytorch_lightning/trainer/supporters.py:187]: This has been adressed I think, but github doesn't seem to notice.

File: pytorch_lightning/trainer/supporters.py [hunk 8]
```diff
@@ -177,3 +179,126 @@ def to_disk(self) -> None:
             # Write predictions for current file to disk
             with fs.open(filepath, "wb") as fp:
                 torch.save(outputs, fp)
+
+
+class AverageValueMeter:
+    '''
+    This class is taken from https://github.com/pytorch/tnt/blob/master/torchnet/meter/averagevaluemeter.py#L6
+    '''
```
Comment (Borda) [inline pytorch_lightning/trainer/supporters.py:187]: @MikkelAntonsen I see it here, as you moved it here in 530bddbd10c82d3adcf3931c1b24acb389255163

**Comment by Borda** [comment]
@tchaton  there won't be any more releases in 1.1.x so I have changed your destination branch to feat 1.2
pls, there is no need to close this PR, just rebase on the target branch, or maybe easier just cherry-pick the commit (@kaushikb11 will help you if needed)

**Comment by edenlightning** [comment]
@tchaton what about this fix?

**Comment by MikkelAntonsen** [comment]
> 
> 
> @tchaton what about this fix?

Hey, things have been a bit hectic for me lately and I've dropped the ball on this one. I'll try to fix the conflicts and failing tests this weekend:)

**Comment by MikkelAntonsen** [comment]
@tchaton I think this is ready for review now. I looked into the failing tpu tests but it is not super obvious to me what's wrong.

**Comment by Borda** [comment]
@tchaton any update here?
